src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,932 @@
+                      _openslide_format_double(mmpp * 1000));
+  g_hash_table_insert(osr->properties,
+                      g_strdup(OPENSLIDE_PROPERTY_NAME_MPP_Y),
+                      _openslide_format_double(mmpp * 1000));
+}
+
+static void file_hash(const struct dicom_file *f,
+                      struct _openslide_hash *quickhash1) {
+  uint32_t n = dcm_dataset_count(f->metadata);
+  for(uint32_t i = 0; i < n; i++) {
+    DcmElement *element = dcm_dataset_get(NULL, f->metadata, i);
+    if (element) {
+      uint32_t tag = dcm_element_get_tag(element);
+      _openslide_hash_data(quickhash1, &tag, sizeof(tag));
+      // need to get the vr class
+      //_openslide_hash_data(quickhash1, data, datalen);
+    }
+  }
+}
+
+static bool dicom_open(openslide_t *osr, 
+                       const char *filename,
+                       struct _openslide_tifflike *tl G_GNUC_UNUSED,
+                       struct _openslide_hash *quickhash1, 
+                       GError **err) {
+  g_autofree char *dirname = g_path_get_dirname(filename);
+
+  g_autoptr(_openslide_dir) dir = _openslide_dir_open(dirname, err);
+  if (!dir) {
+    return false;
+  }
+
+  g_autoptr(GHashTable) dicom_file_hash =
+    g_hash_table_new_full(g_str_hash, 
+                          g_str_equal, 
+                          g_free,
+                          (GDestroyNotify) dicom_file_destroy);
+
+  // open all DICOM files that look like parts of a slide image and 
+  // get the file metadata
+  const char *name;
+  while ((name = _openslide_dir_next(dir))) {
+    g_autofree char *filename = g_build_path("/", dirname, name, NULL);
+
+    g_debug("trying to open: %s ...", filename);
+    GError *local_error = NULL;
+    struct dicom_file *f = dicom_file_new(filename, &local_error);
+    if (!f) {
+      g_debug("open failed: %s", local_error->message);
+      g_error_free(local_error);
+      continue;
+    }
+
+    g_hash_table_insert(dicom_file_hash, g_steal_pointer(&filename), f);
+  }
+
+  g_debug("found WSI DICOM files:");
+  g_hash_table_foreach(dicom_file_hash, print_file_from_hash, NULL);
+
+  // the filename we were passed should be one of these WSI files ... get the
+  // slide id from that
+  struct dicom_file *start = g_hash_table_lookup(dicom_file_hash, filename); 
+  if (!start) {
+    g_set_error(err, OPENSLIDE_ERROR, OPENSLIDE_ERROR_FAILED, 
+        "%s is not a DICOM WSI", filename);
+    return false;
+  }
+
+  if (!read_whole_file(start)) { 
+    return NULL;
+  }
+
+  const char *slide_id;
+  if (!get_tag_str(start->metadata, SeriesInstanceUID, 0, &slide_id)) {
+    return false;
+  }
+
+  // throw away all files which don't have this slide_id
+  g_hash_table_foreach_remove(dicom_file_hash, 
+                              remove_bad_slide_id, 
+                              (char *) slide_id);
+
+  // pull out the subset of DICOM files that look like pyramid levels
+  g_autoptr(GHashTable) level_hash =
```

Comment:
That's a bit awkward since the passed-in filename can be either an associated image or a level or perhaps neither but part of the same slide. You end up needing some confusing extra ifs.

I trimmed it down to just a hash for the matching slide ids and a ptrarray for the levels, is that enough?

Reply:
I think I see what you mean about the extra ifs, but that can be solved with an "add file" function.  I think that's less confusing than calling out to separate functions when iterating.

(And actually, we don't need a separate array for associated images, since we can insert directly into `osr->associated_images` as we go.)

I'm imagining the logic would look roughly like this (high-level pseudocode, not meant literally):

```
func add_file(osr, levels, file) {
  if is_level(file) {
    levels += file
  } else if is_associated(file) {
    osr.associated_images[associated_image_name(file)] = file
  }
}

func dicom_open(osr, filename) {
  base_file = open(filename)
  slide_id = slide_id_of(base_file)
  levels = []
  add_file(osr, levels, base_file)
  foreach file in walkdir(dirname(filename)) {
    // if this check is too hard, we could skip the add_file() above,
    // and maybe inline add_file() below
    if file == base_file {
      continue
    }
    cur_slide_id = slide_id_of(file)
    if cur_slide_id != slide_id {
      continue
    }
    add_file(osr, levels, file)
  }
  sort(levels)
  ...
}
```

Reply:
Is it worth polishing this PR too much in this cycle? We're going to be going over this same code in at least two more PRs within a few weeks. A lot of this work now will have to be redone anyway.

Reply:
Oooof, ok, reorganised as requested.